Tweak: Introduce image loading optimization like WP Core

kt-12 opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [x] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [x] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

kt-12 committed:
LoadingOptimization Module

kt-12 committed:
change filter

kt-12 committed:
cs fixes

kt-12 committed:
remove filter and replace with variables

kt-12 committed:
add_action to filter & add early exit of logic

kt-12 committed:
remove unwanted complexity and also disable core lazy loading and fetchpriority

kt-12 committed:
update codebase

kt-12 committed:
introduce new action to know footer has ended.

kt-12 committed:
filter header and footer

kt-12 committed:
reusing function

kt-12 committed:
remove extra line

kt-12 committed:
remove defaults

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
add empty string

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
fix cbf issue and update optimization code.

kt-12 committed:
change function name to more relevant context.

kt-12 committed:
optimize code further, don't run filter if the image was already loaded.

kt-12 committed:
remove unwanted optimization

kt-12 committed:
Docblock + optimization + fixing header logic

kt-12 committed:
remove the addition hook

kt-12 committed:
remove additional comment sign

kt-12 committed:
change priority to 0

kt-12 committed:
doc block for variables

kt-12 committed:
rename function

kt-12 committed:
update doc block

kt-12 committed:
Enable fetchpriority for is_admin too

kt-12 committed:
test fetch priority

kt-12 committed:
buffer callback check before flushing

kt-12 committed:
use empty instead of isset

kt-12 committed:
remove flush function

kt-12 committed:
Update modules/image-loading-optimization/module.php

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
update logic with preg_replace_callback

kt-12 committed:
Merge branch 'feature/img-lo-wp-logic-to-elementor' of https://github.com/kt-12/elementor into feature/img-lo-wp-logic-to-elementor

kt-12 committed:
test with buffer issue

kt-12 committed:
remove print

kt-12 committed:
path issue fix

kt-12 committed:
handling buffer issue for template.

kt-12 committed:
buffer fix

kt-12 committed:
Merge pull request #2 from kt-12/feature/temp-test

Fix test case with buffer issue.

kt-12 committed:
remove additional function

Co-authored-by: Weston Ruter <westonruter@google.com>

kt-12 committed:
syntax ending issue fix

kt-12 committed:
Add experimental setting

kt-12 committed:
remove extra line

kt-12 committed:
add todo

kt-12 committed:
remove unused namespaces

kt-12 committed:
remove leftover action

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elementor/elementor?pullRequest=23631) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/elementor/elementor?pullRequest=23631) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elementor/elementor?pullRequest=23631) it.</sub>

joemcgill commented:
This PR is an initial implementation of what is described as step 1, **Apply default optimization strategies during content rendering**, in the following proposal:

[Proposal: Better Image Loading Optimization Module](https://github.com/orgs/elementor/discussions/23621)

rami-elementor commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,339 @@
+	 * @param string $content Content to be analyzed.
+	 * @return string Content with optimized images.
+	 */
+	private function filter_images( $content ) {
+		return preg_replace_callback(
+			'/<img\s[^>]+>/',
+			function ( $matches ) {
+				return $this->loading_optimization_image( $matches[0] );
+			},
+			$content
+		);
+	}
```
Hi,

WordPress doesn't know which media file you are using, that is why you need to use regex.

Elementor, on the other hand, has several methods that handle images.

Have you considered hooking to the `elementor/image_size/get_attachment_image_html` [filter](https://github.com/elementor/elementor/blob/main/includes/controls/groups/image-size.php#L137)? This hooks accepts all the relevant data and generates the attachment HTML code.

joemcgill replied:
Hi @rami-elementor! Thanks for pointing this out. 

In order to accurately apply the correct optimizations to each image, we need to know what order they appear in the rendered HTML, which is why running `preg_replace` on all the rendered HTML after all the markup has been rendered is more reliable. In my testing, it doesn't look like all image are run through this filter. For example, the basic image gallery relies on [the gallery shortcode](https://github.com/elementor/elementor/blob/main/includes/widgets/image-gallery.php#L391), which doesn't seem like it will get passed through this filter.

rami-elementor replied:
Ok, thank you for the clarification.

arielk committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
Adjust formatting and CS

joemcgill committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
PHPUnit: Add group annotations

This allows unit tests for the specific image loading optimization module to be run by passing `--group image-loading-optimization` to the CLI.

joemcgill committed:
PHPUnit: Add setup for shared fixtures for image markup

joemcgill committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

TzviRabinovitch commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+			 * interpreted as in viewport (realistically only "eager" or `false`
+			 * to force-omit the attribute are other potential values).
+			 */
+			if ( 'lazy' === $attr['loading'] ) {
```
maybe reduce nesting to improve readability

louiswol94 replied:
@TzviRabinovitch Looks like this code was copied from WP core. I would rather keep it as close to core as possible and just add a comment like `// Copied from wp-includes/media.php`. WDYT?

TzviRabinovitch commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+			 * the most important in-viewport image should have `fetchpriority` set
+			 * to "high".
+			 */
+			if ( false === $maybe_in_viewport ) {
```
maybe reduce nesting to improve readability

joemcgill replied:
This is similarly taken from the implantation in core. Would you prefer it stay consistent as discussed in https://github.com/elementor/elementor/pull/23631/files#r1337317133?

TzviRabinovitch commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+		}
+
+		// Logic to handle a `fetchpriority` attribute that is already provided.
+		if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
```
maybe extract to with is_high_fetch_priority() -> make it more readable

louiswol94 replied:
See my question above. @TzviRabinovitch

TzviRabinovitch commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+		}
+
+		if ( null === $maybe_in_viewport ) {
+			//@todo: Only run this functionality for frontend like in wp core.
```
why are we adding todo? who is going to do it and when? is there a task?

shilo-ey replied:
@TzviRabinovitch I think if this todo is a must - we should create the task. @kt-12 please let me know if it's relevant

joemcgill replied:
Good catch. I'll take another look, but I believe this was to work around an issue with the Unit test suite always running in the admin context, which was causing an early return here.

joemcgill replied:
Addressed in [6e26a95](https://github.com/elementor/elementor/pull/23631/commits/6e26a955605265ae922790bdece4f1e4afc75689)

louiswol94 replied:
@joemcgill If I understand correctly, should this logic only run in the front end? If so, can you check if `Plugin::$instance->editor->is_edit_mode()` works better?

hein-obox commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+			'name' => static::EXPERIMENT_NAME,
+			'title' => esc_html__( 'Optimize Image Loading', 'elementor' ),
+			'tag' => esc_html__( 'Performance', 'elementor' ),
+			'description' => esc_html__( 'Applying `fetchpriority="high"` on LCP image and lazy loading on images below the fold.', 'elementor' ),
```
Could this description be more clear? I find it a bit cryptic now.

shilo-ey replied:
Applying `fetchpriority="high"` on LCP image and lazy loading on images below the fold to improve performance scores.

@kt-12

shilo-ey replied:
This should be "Beta" and set to new sites from v3.17 (min version)

See references: https://github.com/elementor/elementor/pull/23876/files#diff-719a36795ac3daf2c97e2d1a89200ede89ae178f95ad1b85cb0187bd47d4409dR150

@kt-12

joemcgill replied:
Thanks for these suggestions, will do.

louiswol94 committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

louiswol94 commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,71 @@
+	 */
+	private function close_and_print_open_buffer() {
+		$buffer_status = ob_get_status();
+		if( ! empty( $buffer_status ) && str_contains( $buffer_status['name'], 'ImageLoadingOptimization\Module::handle_buffer_content') ) {
```
`str_contains()` is added in PHP8, but our PHPUnit tests will run on PHP7 as well, so this might be a possible point of failure.

joemcgill replied:
WP Core has a polyfill for this functionality, since WP 5.9, See: https://core.trac.wordpress.org/ticket/49652. Is that ok?

louiswol94 replied:
Ah, nice! 👍

louiswol94 commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,71 @@
+	 * This is to ensure buffer started in footer is closed.
+	 */
+	private function close_and_print_open_buffer() {
+		$buffer_status = ob_get_status();
+		if( ! empty( $buffer_status ) && str_contains( $buffer_status['name'], 'ImageLoadingOptimization\Module::handle_buffer_content') ) {
```
```suggestion
		$buffer_status = ob_get_status();

		if ( ! empty( $buffer_status ) && str_contains( $buffer_status['name'], 'ImageLoadingOptimization\Module::handle_buffer_content') ) {
```

louiswol94 commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+	 * @return array Loading optimization attributes.
+	 */
+	private function get_loading_optimization_attributes( $attr ) {
+		$loading_attrs = array();
```
```suggestion
		$loading_attrs = [];
```

louiswol94 commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+
+		// Stop wp core logic.
+		add_action( 'init', [ $this, 'stop_core_fetchpriority_high_logic' ] );
+		add_filter( 'wp_lazy_loading_enabled', '__return_false' );
```
If we explicitly set this to false here, I am wondering if there might be any negative side-effects for 3rd-party plugins that use this?
![Screenshot 2023-09-26 at 16 24 10](https://github.com/elementor/elementor/assets/20996103/2b0c8cb4-8c18-43ca-9f63-712539e166d2)

joemcgill replied:
Good question. We disable it here because we're completely replacing the functionality, but let me consider a way that we could support this core flag and still over load the functionality that depends on it.

louiswol94 commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+	/**
+	 * @var int Minimum square-pixels threshold.
+	 */
+	private $min_priority_img_pixels = 50000;
```
Should there be a filter for this value (as there is in WP)?

joemcgill replied:
Yes, we should add one here. Good catch

louiswol94 commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+		}
+
+		$optimization_attrs = $this->get_loading_optimization_attributes(
+			array(
+				'width'         => $width,
+				'height'        => $height,
+				'loading'       => $loading_val,
+				'fetchpriority' => $fetchpriority_val,
+			)
```
```suggestion
			[
				'width'         => $width,
				'height'        => $height,
				'loading'       => $loading_val,
				'fetchpriority' => $fetchpriority_val,
			]
```
Short array syntax please 😃

louiswol94 commented:
@kt-12 Please check failing PHPUnit tests:

```
There were 2 failures:

1) Elementor\Tests\Phpunit\Elementor\Modules\Image_Loading_Optimization\Elementor_Image_Loading_Optimization_Test_Module::test_loading_optimization_without_logo with data set #0 ('elementor_canvas')
Unexpected deprecation notice for print_emoji_styles.
Function print_emoji_styles is deprecated since version 6.4.0! Use wp_enqueue_emoji_styles instead.
Failed asserting that an array is empty.

/tmp/wordpress-tests-lib/includes/abstract-testcase.php:637
/tmp/wordpress-tests-lib/includes/abstract-testcase.php:649
/home/runner/work/elementor/elementor/vendor/yoast/phpunit-polyfills/src/TestCases/TestCasePHPUnitLte7.php:96
phpvfscomposer:///home/runner/work/elementor/elementor/vendor/phpunit/phpunit/phpunit:60

2) Elementor\Tests\Phpunit\Elementor\Modules\Image_Loading_Optimization\Elementor_Image_Loading_Optimization_Test_Module::test_loading_optimization_without_logo with data set #1 ('elementor_header_footer')
Unexpected deprecation notice for print_emoji_styles.
Function print_emoji_styles is deprecated since version 6.4.0! Use wp_enqueue_emoji_styles instead.
Failed asserting that an array is empty.

/tmp/wordpress-tests-lib/includes/abstract-testcase.php:637
/tmp/wordpress-tests-lib/includes/abstract-testcase.php:649
/home/runner/work/elementor/elementor/vendor/yoast/phpunit-polyfills/src/TestCases/TestCasePHPUnitLte7.php:96
phpvfscomposer:///home/runner/work/elementor/elementor/vendor/phpunit/phpunit/phpunit:60
```

rami-elementor commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+			'name' => static::EXPERIMENT_NAME,
+			'title' => esc_html__( 'Optimize Image Loading', 'elementor' ),
+			'tag' => esc_html__( 'Performance', 'elementor' ),
+			'description' => esc_html__( 'Applying `fetchpriority="high"` on LCP image and lazy loading on images below the fold.', 'elementor' ),
```
```suggestion
			'description' => sprintf(
				/* translators: 1: fetchpriority attribute, 2: lazy loading attribute. */
				esc_html__( 'Applying %s on LCP image and %s on images below the fold.', 'elementor' ),
				'<code>fetchpriority="high"</code>',
				'<code>loading="lazy"</code>'
			),
```

joemcgill committed:
Use short array syntax

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

joemcgill committed:
Improve formatting of translated description

Co-authored-by: Rami Yushuvaev <92088692+rami-elementor@users.noreply.github.com>

joemcgill committed:
Add line break for readability

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

joemcgill reviewed (commented):
Thanks for all the feedback everyone. I've left a few questions below and will work through the feedback shortly.

joemcgill committed:
Abstract conditional logic to a variable for readability

joemcgill committed:
Update module description and release status.

joemcgill committed:
Allow minimum pixel threshold to be filterable

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,346 @@
+	 * @return array Updated loading optimization attributes for the element.
+	 */
+	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ) {
+		if ( isset( $attr['fetchpriority'] ) ) {
+			if ( 'high' === $attr['fetchpriority'] ) {
```
This one is a bit tricky, because we need to do two different things based on whether the attribute exists and if it has already been set to 'high' so any attributes added intentionally are respected and not overwritten here. That's why this is split into a nested set of conditionals.

louiswol94 replied:
@joemcgill Is this not functionally equivalent? 🤔

```
if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
        $loading_attrs['fetchpriority'] = 'high';
        $this->high_priority_element_flag( false );
        return $loading_attrs;
}
```

louiswol94 replied:
I see 👍

joemcgill commented:
@louiswol94 regarding a few of your comments above:

> Please check failing PHPUnit tests

Those test failures are related to [a recent change](https://core.trac.wordpress.org/changeset/56682) in WP for version 6.4 that are still likely to change during the beta period for this WordPress release. It is not something that effects this functionality. Even so, is there something we need to do to make it possible to get the checks for this PR to run so test status is visible via CI?

> I guess scanning the entire `$content` is the only sure fire way to make sure we catch all the images though

That's correct. We need to be able to make sure images are processed in the same order that they show up in the rendered markup. This is very similar to what WordPress is already doing to dynamically optimize the image markup, but we don't require any additional DB queries, so it's not too costly. I believe the `handle_buffer_content` method should only run twice: once for the theme's header and again for the footer (and only when the site is not using the canvas template). Still, it will be nice to get some A/B benchmarks for this, which I plan to supply this week.

joemcgill commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -12,6 +12,47 @@
+	/**
+	 * Setup shared attachments.
+	 */
+	public static function set_up_before_class() {
```
In order to add tests for multiple layout use cases, it would be nice to have a set of test images that we can set up during these tests. Do you all have any preferred way of adding test images to the test suite? I've set up the process here that should make creating these attachments easier during a test run, but didn't add any more image files until I could get feedback from you.

louiswol94 replied:
@joemcgill I am sure you can add more images to the resources folder 👍

joemcgill committed:
Fix is_admin check in unit tests

louiswol94 commented:
> This is similarly taken from the implementation in core. Would you prefer it stay consistent as discussed in https://github.com/elementor/elementor/pull/23631/files#r1337317133?
@joemcgill Yes please. You can just add the comment as discussed. Thanks!

joemcgill commented:
@shilo-ey since you had asked about running some performance metrics, I've taken an initial pass at providing some metrics below. Happy to answer any questions about these, and would be really curious to see any additional performance metrics you are able to share based on your test processes.

# Benchmarks

These benchmark some key performance data of [this PR](https://github.com/elementor/elementor/pull/23631) against the main branch when running in WordPress versions 6.3.1 and 6.2.2, respectively. The CLI tools from the [WPP Research](https://github.com/GoogleChromeLabs/wpp-research/tree/main/cli#benchmark-web-vitals) repo have been used to benchmark both server timing and Core Web Vitals (CWV) metrics. For this specific test, a simple use case was created that is known to not work as expected with the logic in WP Core as of 6.3.1, which contains an image widget followed by an image gallery widget containing three images.

## WordPress 6.3.1 (latest release)

### Server Timing

Interestingly, this shows some possible server timing improvement, possibly due to the fact that we are disabling Core functionality and replacing it with a more performant alternative. However, these improvements are also just as likely to be due to some expected variance and not to be overly relied upon.

**Main**

```
╔═════════════════════╤══════════════════════════════════╗
║ URL                 │ http://localhost:8888/test-post/ ║
╟─────────────────────┼──────────────────────────────────╢
║ Success Rate        │ 100%                             ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p10) │ 37.4                             ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p25) │ 37.66                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p50) │ 40.05                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p75) │ 41.93                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p90) │ 43.86                            ║
╚═════════════════════╧══════════════════════════════════╝
```

**PR**

```
╔═════════════════════╤══════════════════════════════════╗
║ URL                 │ http://localhost:8888/test-post/ ║
╟─────────────────────┼──────────────────────────────────╢
║ Success Rate        │ 100%                             ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p10) │ 34.41                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p25) │ 34.83                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p50) │ 35.52                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p75) │ 37.45                            ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p90) │ 39.71                            ║
╚═════════════════════╧══════════════════════════════════╝
```

### Core Web Vitals

The improvements to LCP scores are not all that significant here, since the enhancements to LCP detection shipped in WP 6.3.1 already do a pretty good job of avoiding the worst outcomes in this test. However, this does improve the expected markup and avoids a bug with shortcodes that would expect to negatively impact Elementor sites running on WP 6.3.x versions.

**Main**

```
╔════════════════╤══════════════════════════════════╗
║ URL            │ http://localhost:8888/test-post/ ║
╟────────────────┼──────────────────────────────────╢
║ Success Rate   │ 100%                             ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p10)      │ 99.66                            ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p25)      │ 103.27                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p50)      │ 106.2                            ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p75)      │ 110.85                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p90)      │ 144.77                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p10)      │ 99.66                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p25)      │ 103.27                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p50)      │ 106.2                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p75)      │ 110.85                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p90)      │ 144.77                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p10)     │ 58.89                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p25)     │ 61.35                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p50)     │ 62.75                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p75)     │ 67.65                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p90)     │ 97.56                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p10) │ 40.79                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p25) │ 41.27                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p50) │ 42.9                             ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p75) │ 44.55                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p90) │ 46.49                            ║
╚════════════════╧══════════════════════════════════╝
```

**PR**

```
╔════════════════╤══════════════════════════════════╗
║ URL            │ http://localhost:8888/test-post/ ║
╟────────────────┼──────────────────────────────────╢
║ Success Rate   │ 100%                             ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p10)      │ 100.63                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p25)      │ 101.9                            ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p50)      │ 104.4                            ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p75)      │ 109.38                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p90)      │ 116.22                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p10)      │ 100.63                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p25)      │ 101.9                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p50)      │ 104.4                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p75)      │ 109.38                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p90)      │ 116.22                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p10)     │ 58.98                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p25)     │ 60.12                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p50)     │ 62.8                             ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p75)     │ 65.03                            ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p90)     │ 68.62                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p10) │ 40.76                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p25) │ 41.62                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p50) │ 42.55                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p75) │ 43.92                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p90) │ 46.01                            ║
╚════════════════╧══════════════════════════════════╝
```

## WP 6.2.2 (Previous Release)

### Server Timing

As expected, this PR has little to no impact on server response times, as the purpose is to improve front end rendering.

**Main**

```
╔═════════════════════╤══════════════════════════════════╗
║ URL                 │ http://localhost:8888/test-post/ ║
╟─────────────────────┼──────────────────────────────────╢
║ Success Rate        │ 100%                             ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p10) │ 114.69                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p25) │ 115.95                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p50) │ 117.92                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p75) │ 119.99                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p90) │ 134.48                           ║
╚═════════════════════╧══════════════════════════════════╝
```

**PR**

```
╔═════════════════════╤══════════════════════════════════╗
║ URL                 │ http://localhost:8888/test-post/ ║
╟─────────────────────┼──────────────────────────────────╢
║ Success Rate        │ 100%                             ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p10) │ 115.14                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p25) │ 116.36                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p50) │ 117.73                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p75) │ 118.59                           ║
╟─────────────────────┼──────────────────────────────────╢
║ Response Time (p90) │ 122.51                           ║
╚═════════════════════╧══════════════════════════════════╝
```

### Core Web Vitals

The key LCP metrics that are being targetted (LCP, and LCP-TTFB) both show improvement. This makes sense because in our test cases, without the PR the main image is being lazy loaded and no fetchpriority attribute is present. With the PR applied, the first image is no longer lazy loaded and the first image has `fetchpriority="high"` added.

**Main**

```
╔════════════════╤══════════════════════════════════╗
║ URL            │ http://localhost:8888/test-post/ ║
╟────────────────┼──────────────────────────────────╢
║ Success Rate   │ 100%                             ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p10)      │ 297.93                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p25)      │ 303.62                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p50)      │ 318.55                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p75)      │ 356.2                            ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p90)      │ 407.64                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p10)      │ 297.93                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p25)      │ 303.62                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p50)      │ 318.55                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p75)      │ 356.2                            ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p90)      │ 407.64                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p10)     │ 149.81                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p25)     │ 151.57                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p50)     │ 156.45                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p75)     │ 167.35                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p90)     │ 179.06                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p10) │ 148.27                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p25) │ 149.42                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p50) │ 158.45                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p75) │ 179.17                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p90) │ 223.24                           ║
╚════════════════╧══════════════════════════════════╝
```

**PR**

```
╔════════════════╤══════════════════════════════════╗
║ URL            │ http://localhost:8888/test-post/ ║
╟────────────────┼──────────────────────────────────╢
║ Success Rate   │ 100%                             ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p10)      │ 296.49                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p25)      │ 306.05                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p50)      │ 313.85                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p75)      │ 338.97                           ║
╟────────────────┼──────────────────────────────────╢
║ FCP (p90)      │ 362.46                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p10)      │ 296.49                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p25)      │ 306.05                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p50)      │ 313.85                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p75)      │ 338.97                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP (p90)      │ 362.46                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p10)     │ 150.69                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p25)     │ 154.13                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p50)     │ 159.45                           ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p75)     │ 172                              ║
╟────────────────┼──────────────────────────────────╢
║ TTFB (p90)     │ 178.38                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p10) │ 148.17                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p25) │ 151.25                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p50) │ 154.15                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p75) │ 165.12                           ║
╟────────────────┼──────────────────────────────────╢
║ LCP-TTFB (p90) │ 178.02                           ║
╚════════════════╧══════════════════════════════════╝
```

joemcgill committed:
Add inline docs about copied logic

louiswol94 commented:
> Those test failures are related to [a recent change](https://core.trac.wordpress.org/changeset/56682) in WP for version 6.4 that are still likely to change during the beta period for this WordPress release. It is not something that effects this functionality. Even so, is there something we need to do to make it possible to get the checks for this PR to run so test status is visible via CI?

@joemcgill I need to manually approve these workflow runs https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks - I think for now just ping me when you think we can run it again? If we run into too many bottlenecks because of this we'll make a plan.

louiswol94 commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,119 @@
+		$output = ob_get_clean();
+
+		$expected = '<p><img fetchpriority="high" decoding="async" width="800" height="530" src="featured_image.jpg" /><img decoding="async" width="640" height="471" src="image_1.jpg" /><img decoding="async" width="800" height="800" src="image_2.jpg" /><img loading="lazy" decoding="async" width="566" height="541" src="image_3.jpg" /><img loading="lazy" decoding="async" width="691" height="1024" src="im ...
+		$this->assertStringContainsString( $expected, $output, "Loading optimization not applied to the content");
```
```suggestion
		$this->assertStringContainsString( $expected, $output, "Loading optimization not applied to the content" );
```

louiswol94 commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,119 @@
+		the_post();
+
+		// get template path from post. If it's not set switch to default single page template.
+		$template_path = locate_template('single.php');
```
```suggestion
		$template_path = locate_template( 'single.php' );
```

louiswol94 commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,119 @@
+		rewind_posts();
+
+		ob_start();
+		load_template($template_path);
```
```suggestion
		load_template( $template_path );
```

shilo-ey commented:
> @shilo-ey since you had asked about running some performance metrics, I've taken an initial pass at providing some metrics below. Happy to answer any questions about these, and would be really curious to see any additional performance metrics you are able to share based on your test processes.
> 
> # Benchmarks
> These benchmark some key performance data of [this PR](https://github.com/elementor/elementor/pull/23631) against the main branch when running in WordPress versions 6.3.1 and 6.2.2, respectively. The CLI tools from the [WPP Research](https://github.com/GoogleChromeLabs/wpp-research/tree/main/cli?rgh-link-date=2023-09-26T22%3A01%3A21Z#benchmark-web-vitals) repo have been used to benchmark both server timing and Core Web Vitals (CWV) metrics. For this specific test, a simple use case was created that is known to not work as expected with the logic in WP Core as of 6.3.1, which contains an image widget followed by an image gallery widget containing three images.
> 
> ## WordPress 6.3.1 (latest release)
> ### Server Timing
> Interestingly, this shows some possible server timing improvement, possibly due to the fact that we are disabling Core functionality and replacing it with a more performant alternative. However, these improvements are also just as likely to be due to some expected variance and not to be overly relied upon.
> 
> **Main**
> 
> ```
> ╔═════════════════════╤══════════════════════════════════╗
> ║ URL                 │ http://localhost:8888/test-post/ ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Success Rate        │ 100%                             ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p10) │ 37.4                             ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p25) │ 37.66                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p50) │ 40.05                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p75) │ 41.93                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p90) │ 43.86                            ║
> ╚═════════════════════╧══════════════════════════════════╝
> ```
> 
> **PR**
> 
> ```
> ╔═════════════════════╤══════════════════════════════════╗
> ║ URL                 │ http://localhost:8888/test-post/ ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Success Rate        │ 100%                             ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p10) │ 34.41                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p25) │ 34.83                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p50) │ 35.52                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p75) │ 37.45                            ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p90) │ 39.71                            ║
> ╚═════════════════════╧══════════════════════════════════╝
> ```
> 
> ### Core Web Vitals
> The improvements to LCP scores are not all that significant here, since the enhancements to LCP detection shipped in WP 6.3.1 already do a pretty good job of avoiding the worst outcomes in this test. However, this does improve the expected markup and avoids a bug with shortcodes that would expect to negatively impact Elementor sites running on WP 6.3.x versions.
> 
> **Main**
> 
> ```
> ╔════════════════╤══════════════════════════════════╗
> ║ URL            │ http://localhost:8888/test-post/ ║
> ╟────────────────┼──────────────────────────────────╢
> ║ Success Rate   │ 100%                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p10)      │ 99.66                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p25)      │ 103.27                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p50)      │ 106.2                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p75)      │ 110.85                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p90)      │ 144.77                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p10)      │ 99.66                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p25)      │ 103.27                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p50)      │ 106.2                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p75)      │ 110.85                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p90)      │ 144.77                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p10)     │ 58.89                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p25)     │ 61.35                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p50)     │ 62.75                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p75)     │ 67.65                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p90)     │ 97.56                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p10) │ 40.79                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p25) │ 41.27                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p50) │ 42.9                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p75) │ 44.55                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p90) │ 46.49                            ║
> ╚════════════════╧══════════════════════════════════╝
> ```
> 
> **PR**
> 
> ```
> ╔════════════════╤══════════════════════════════════╗
> ║ URL            │ http://localhost:8888/test-post/ ║
> ╟────────────────┼──────────────────────────────────╢
> ║ Success Rate   │ 100%                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p10)      │ 100.63                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p25)      │ 101.9                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p50)      │ 104.4                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p75)      │ 109.38                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p90)      │ 116.22                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p10)      │ 100.63                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p25)      │ 101.9                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p50)      │ 104.4                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p75)      │ 109.38                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p90)      │ 116.22                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p10)     │ 58.98                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p25)     │ 60.12                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p50)     │ 62.8                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p75)     │ 65.03                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p90)     │ 68.62                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p10) │ 40.76                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p25) │ 41.62                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p50) │ 42.55                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p75) │ 43.92                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p90) │ 46.01                            ║
> ╚════════════════╧══════════════════════════════════╝
> ```
> 
> ## WP 6.2.2 (Previous Release)
> ### Server Timing
> As expected, this PR has little to no impact on server response times, as the purpose is to improve front end rendering.
> 
> **Main**
> 
> ```
> ╔═════════════════════╤══════════════════════════════════╗
> ║ URL                 │ http://localhost:8888/test-post/ ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Success Rate        │ 100%                             ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p10) │ 114.69                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p25) │ 115.95                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p50) │ 117.92                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p75) │ 119.99                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p90) │ 134.48                           ║
> ╚═════════════════════╧══════════════════════════════════╝
> ```
> 
> **PR**
> 
> ```
> ╔═════════════════════╤══════════════════════════════════╗
> ║ URL                 │ http://localhost:8888/test-post/ ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Success Rate        │ 100%                             ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p10) │ 115.14                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p25) │ 116.36                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p50) │ 117.73                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p75) │ 118.59                           ║
> ╟─────────────────────┼──────────────────────────────────╢
> ║ Response Time (p90) │ 122.51                           ║
> ╚═════════════════════╧══════════════════════════════════╝
> ```
> 
> ### Core Web Vitals
> The key LCP metrics that are being targetted (LCP, and LCP-TTFB) both show improvement. This makes sense because in our test cases, without the PR the main image is being lazy loaded and no fetchpriority attribute is present. With the PR applied, the first image is no longer lazy loaded and the first image has `fetchpriority="high"` added.
> 
> **Main**
> 
> ```
> ╔════════════════╤══════════════════════════════════╗
> ║ URL            │ http://localhost:8888/test-post/ ║
> ╟────────────────┼──────────────────────────────────╢
> ║ Success Rate   │ 100%                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p10)      │ 297.93                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p25)      │ 303.62                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p50)      │ 318.55                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p75)      │ 356.2                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p90)      │ 407.64                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p10)      │ 297.93                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p25)      │ 303.62                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p50)      │ 318.55                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p75)      │ 356.2                            ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p90)      │ 407.64                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p10)     │ 149.81                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p25)     │ 151.57                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p50)     │ 156.45                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p75)     │ 167.35                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p90)     │ 179.06                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p10) │ 148.27                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p25) │ 149.42                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p50) │ 158.45                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p75) │ 179.17                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p90) │ 223.24                           ║
> ╚════════════════╧══════════════════════════════════╝
> ```
> 
> **PR**
> 
> ```
> ╔════════════════╤══════════════════════════════════╗
> ║ URL            │ http://localhost:8888/test-post/ ║
> ╟────────────────┼──────────────────────────────────╢
> ║ Success Rate   │ 100%                             ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p10)      │ 296.49                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p25)      │ 306.05                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p50)      │ 313.85                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p75)      │ 338.97                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ FCP (p90)      │ 362.46                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p10)      │ 296.49                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p25)      │ 306.05                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p50)      │ 313.85                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p75)      │ 338.97                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP (p90)      │ 362.46                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p10)     │ 150.69                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p25)     │ 154.13                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p50)     │ 159.45                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p75)     │ 172                              ║
> ╟────────────────┼──────────────────────────────────╢
> ║ TTFB (p90)     │ 178.38                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p10) │ 148.17                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p25) │ 151.25                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p50) │ 154.15                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p75) │ 165.12                           ║
> ╟────────────────┼──────────────────────────────────╢
> ║ LCP-TTFB (p90) │ 178.02                           ║
> ╚════════════════╧══════════════════════════════════╝
> ```

Really good job @joemcgill , thank you!

As for the metrics data - I think there's some glitch in LCP and FCP data- it seems to be repeating itself (LCP and FCP have the same values in each environment). Do you know why?

louiswol94 committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
CS Fixes from code review

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

louiswol94 commented:
@joemcgill I fixed the `print_emoji_styles` error by temporarily disabling our PHPUnit tests for WP Nightly.

joemcgill commented:
Thanks! I'm looking into the PHPUnit errors in 6.0 and looks like I need to just make some tweaks to the expected output, based on some changes to the image markup that began after WP 6.0. Should have that wrapped up today.

joemcgill committed:
Improve unit teststo be less brittle

joemcgill commented:
@louiswol94 I've pushed a change to the unit tests that should make them pass against WP 6.0. Would you mind approving the workflows again to confirm?

louiswol94 commented:
> @louiswol94 I've pushed a change to the unit tests that should make them pass against WP 6.0. Would you mind approving the workflows again to confirm?

LGTM! Tests passed 😃

louiswol94 commented:
@joemcgill @kt-12 Thank you for the contribution!
